### ISSUE TITLE
transport client created by plugin cannot load plugin classes

### ISSUE THREAD

**Issue by jaymode** [elastic/elasticsearch#14684]
In 2.0, if a plugin has code that creates a transport client and specifies the plugin using the `plugin.types` setting, the transport client cannot be created due to a ClassNotFoundException.

```java
TransportClient transportClient = TransportClient.builder()
                    .settings(Settings.builder()
                            .putArray("plugin.types", MyPlugin.class.getName()
                            ...)
                    .build();
```

results in:

```
ElasticsearchException[Failed to load plugin class [org.elasticsearch.MyPlugin]]; nested: ClassNotFoundException[org.elasticsearch.MyPlugin];
	at org.elasticsearch.plugins.PluginsService.loadPlugin(PluginsService.java:398)
	at org.elasticsearch.plugins.PluginsService.<init>(PluginsService.java:101)
	at org.elasticsearch.client.transport.TransportClient$Builder.build(TransportClient.java:114)
```

This fails since the classloader used by the `PluginsService` does not know about the plugin jars and therefore plugin classes.

I've verified that this is not a problem in 2.1 since the plugin loading mechanism has been changed to pass in the actual plugin classes. Its worth discussing if/how this should be fixed in 2.0.x

**Comment by rjernst** [comment]
The fix that is in 2.1 was complex. I don't think we should make such a complicated change in a bugfix release.

**Comment by bleskes** [comment]
@rjernst can you point at the commit that fixed this in 2.1? my search foo fails me.

**Comment by rjernst** [comment]
It was the refactoring to remove plugin.types

**Comment by rjernst** [comment]
@bleskes See #13055

**Comment by jaymode** [comment]
Closing since this is fixed in 2.1 and the change is too big for a bugfix release.
